#162948860 Like and Dislike Comments

KMaina opened the pull request:
#### What does this PR do?
Allow for users to like and dislike comments
#### Description of Task to be completed?
Like a comment made 

`PUT api/articles/<slug>/comments/<int:id>/like`

Dislike a comment made

`PUT api/articles/<slug>/comments/<int:id>/dislike`

#### How should this be manually tested?

- Clone the repo locally

- cd ah-django-unchained
- git checkout ft-like-comment-162948860
- get started with the app as from the README section. Follow the sample .env file supplied to obtain the right variables
- create a user and their bio
- create an article as required.
- create a comment for an article
- to like an article use `PUT api/articles/<slug>/comments/<int:id>/like`
- to dislike an article use `PUT api/articles/<slug>/comments/<int:id>/dislike`

#### What are the relevant pivotal tracker stories?
[#162948847](https://www.pivotaltracker.com/story/show/162948847)

KMaina committed:
feat(like comments): add migrations for database changes

[Delivers #162948847]

KMaina committed:
feat(like comments): add changes made to the model

[Delivers #162948847]

KMaina committed:
feat(like comments): add changes made in serializer

- add a like and dislike field
- add total count for likes and dislikes

[Delivers #162948847]

KMaina committed:
feat(like comments): add tests for liking and disliking articles

[Delivers #162948847]

KMaina committed:
feat(like comments): add new routes for the view

- add view for liking a comment
- add view for disliking a comments

[Delivers #162948847]

KMaina committed:
feat(like comments): add views for liking and disliking articles

[Delivers #162948847]

KMaina committed:
feat(favorite articles): merge migrations

[Delivers #162948847]

MaggieChege commented on authors/apps/articles/tests/test_comment_like_dislike.py:
```diff
@@ -0,0 +1,176 @@
+from django.urls import reverse
+from rest_framework.views import status
+from rest_framework.test import APITestCase, APIClient
+from authors.apps.articles.models import Article
```
Get rid of this since you are not using it.

KMaina committed:
feat(favorite articles): remove unused import

[Delivers #162948860]

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -390,3 +392,88 @@ def retrieve(self, request, slug, *args, **kwargs):
+            # get an article with the specified slug
+            # and return an exception if the article does not exist
+            try:
+                Article.objects.get(slug=article_slug)
```
This should also check if the article has been deleted

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -390,3 +392,88 @@ def retrieve(self, request, slug, *args, **kwargs):
+                Article.objects.get(slug=article_slug)
+            except ObjectDoesNotExist:
+                raise NotFound("Article does not exist")
+            comment = Comment.objects.get(id=comment_id)
```
This should also check if the comment has been deleted

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -390,3 +392,88 @@ def retrieve(self, request, slug, *args, **kwargs):
+            raise NotFound("A comment from this article does not exist")
+
+        # check if the comment has been disliked by the current user
+        # if the current user has disliked it, remove their id from the row
```
Check your comment

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -390,3 +392,88 @@ def retrieve(self, request, slug, *args, **kwargs):
+            comment.user_id_likes.remove(request.user)
+
+        # if the current user had previously disliked this comment,
+        # then the comment is disliked
```
Check your comment

KMaina committed:
feat(favorite articles): add more test cases

- Test to check if article has been deleted
- Test to check if comment has been deleted

[Delivers #162948860]

KMaina committed:
feat(favorite articles): assert for deletion of articles and comments

[Delivers #162948860]

KMaina committed:
feat(favorite articles): change commented code

[Delivers #162948860]

Georgeygigz commented on authors/apps/articles/tests/test_comment_like_dislike.py:
```diff
@@ -0,0 +1,244 @@
+from django.urls import reverse
+from rest_framework.views import status
+from rest_framework.test import APITestCase, APIClient
+# from authors.apps.articles.models import Article
```
Remove the commented import

Georgeygigz commented on authors/apps/articles/tests/test_comment_like_dislike.py:
```diff
@@ -0,0 +1,244 @@
+class CreateArticles(APITestCase):
+    def setUp(self):
+        self.client = APIClient()
+        self.login_url = reverse('authentication:auth-login')
```
This is not used, kindly consider removing it.

Georgeygigz commented on authors/apps/articles/tests/test_comment_like_dislike.py:
```diff
@@ -0,0 +1,244 @@
+        self.login_url = reverse('authentication:auth-login')
+        self.signup_url = reverse('authentication:auth-register')
+        self.article_listcreate = reverse('articles:articles-listcreate')
+        self.article_favorite = reverse('articles:articles-favorite',
```
ditto

KMaina committed:
feat(favorite articles): remove unused code

[Delivers #162948860]

KMaina committed:
feat(favorite articles): merge and make migrations

[Delivers #162948860]
